server/coordinator/procedure/manager_impl.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+package procedure
+
+import (
+	"context"
+	"sync"
+
+	"github.com/CeresDB/ceresmeta/pkg/log"
+	"github.com/CeresDB/ceresmeta/server/coordinator/eventdispatch"
+	"github.com/pkg/errors"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.uber.org/zap"
+)
+
+const (
+	queueSize         = 10
+	metaListBatchSize = 100
+)
+
+type ManagerImpl struct {
+	// This lock is used to protect the field `procedures`.
+	lock       sync.RWMutex
+	procedures []Procedure
+	running    bool
+
+	storage  Storage
+	dispatch eventdispatch.Dispatch
+
+	procedureQueue chan Procedure
+}
+
+func (m *ManagerImpl) Start(ctx context.Context) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if m.running {
+		log.Warn("cluster manager has already been started")
+		return nil
+	}
+	m.procedureQueue = make(chan Procedure, queueSize)
+	go m.startProcedureWorker(ctx, m.procedureQueue)
+	err := m.retryAll(ctx)
+	if err != nil {
+		return errors.WithMessage(err, "retry all running procedure failed")
+	}
+	return nil
+}
+
+func (m *ManagerImpl) Stop(ctx context.Context) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	close(m.procedureQueue)
+	for _, procedure := range m.procedures {
+		if procedure.State() == StateRunning {
+			err := procedure.Cancel(ctx)
+			log.Error("cancel procedure failed", zap.Error(err), zap.Uint64("procedureID", procedure.ID()))
+			// TODO: consider whether a single procedure cancel failed should return directly.
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *ManagerImpl) Submit(_ context.Context, procedure Procedure) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.procedures = append(m.procedures, procedure)
+	return nil
+}
+
+func (m *ManagerImpl) Cancel(ctx context.Context, procedureID uint64) error {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	for _, procedure := range m.procedures {
+		if procedure.ID() == procedureID {
+			m.removeProcedure(procedure)
```

Comment:
It will cause deadlock here, and I guess we should make the `removeProcedure` can return the removed Procedure, with which we can do cancel here.